spec/mempool/mempool.md

```diff
@@ -0,0 +1,152 @@
+
+**Ledger.**
+We use the standard definition of (BFT) SMR in the context of blockchain, where each process $p$ has a ledger, written $p.ledger$.
+At process $p$, the $i$-th entry of the ledger is denoted $p.ledger[i]$.
+This entry contains either a null value ($\bot$), or a set of transactions, aka., a block.
+The height of the ledger at $p$ is the index of the first null entry; denoted $p.height$.
+Operation $submit(txs, i)$ attempts to write the set of transactions $txs$ to the $i$-th entry of the ledger.
+The (history) variable $p.submitted[i]$ holds all the transactions (if any) submitted by $p$ at height $i$.
+By extension, $p.submitted$ are all the transaction submitted by $p$.
+A transaction is committed when it appears in one of the entries of the ledger.
+We write $p.committed$ the committed transactions at $p$.
+
+As standard, the ledger ensures that:  
+* _(Gap-freedom)_ There is no gap between two entries at a correct process:  
+$\forall i \in 	\mathbb{N}. \forall p \in Correct. \square(p.ledger[i] \neq \bot \implies (i=0 \vee p.ledger[i-1] \neq \bot))$;  
+* _(Agreement)_ No two correct processes have different ledger entries; formally:  
+$\forall i \in 	\mathbb{N}. \forall p,q \in Correct. \square((p.ledger[i] = \bot) \vee (q.ledger[i] = \bot) \vee (p.ledger[i] = q.ledger[i]))$;  
+* _(Validity)_ If some transaction appears at an index $i$ at a correct process, then a process submitted it at that index:  
+$\forall p \in Correct. \exists q \in Processes. \forall i \in 	\mathbb{N}. \square(tx \in p.ledger[i] \implies tx \in \bigcup_q q.submitted[i]$).
+* _(Termination)_ If a correct process submits a block at its current height, eventually its height get incremented:  
+$\forall p \in Correct. \square((h=p.height \wedge p.submitted[h] \neq \varnothing) \implies \lozenge(p.height>h))$  
+
+**Mempool.**
+A mempool is a replicated set of transactions.
+At a process $p$, we write it $p.mempool$.
+We also define $p.hmempool$, the (history) variable that tracks all the transactions ever added to the mempool by process $p$.
+Below, we list the invariants of the mempool (at a correct process).
+
+The mempool is used as an input for the ledger:  
+**INV1.** $\forall tx. \forall p \in Correct. \square(tx \in p.submitted \implies tx \in p.hmempool)$
+
+Committed transactions are not in the mempool:  
+**INV2.** $\forall tx. \forall p \in Correct. \square(tx \in p.committed \implies tx \notin p.mempool)$
+
+In blockchain, a transaction is (or not) valid in a given state.
+That is, a transaction can be valid (or not) at a given height of the ledger.
+To model this, consider a transaction $tx$.
+Let $p.ledger.valid(tx)$ be such a check at the current height of the ledger at process $p$ (ABCI call).
+Our third invariant is that only valid transactions are present in the mempool:  
+**INV3.** $\forall tx, \forall p \in Correct. \square(tx \in p.mempool \implies p.ledger.valid(tx))$
```

Comment:
There is also some atomicity requirement. Whenever a block is appended, a transaction may be invalidated by that. This assumes that p immediately removes tx from the mempool. Again, for the high level spec this is OK, but in the implementation removal of the mempool may be delayed, I guess?

Reply:
Yes, the current implementation is executing the removal asynchronously.